### ISSUE TITLE
how does Horovod do allreduce after each epoch

### ISSUE THREAD

**Issue by windson87** [horovod/horovod#890]
Hi, Could you please kindly clarify some questions from "pytorch_mnist.py" script.

My question is: we do data sampling using torch.utils.data.distributed.DistributedSampler, and then broadcast params from Rank 0 to other GPUs to guarantee every GPU has the same params to start.

But we did it outside the training loop so seems only done once at the beginning of the training. So how does Horovod do the gradient collection and average after each epoch? coz In my understanding, every GPU needs to have the same params for every epoch, so allreduce process should be done per epoch, Am I right on this point?

 2) the other question is, I am now running my own script, similar to mnist script on azureml with horovod support. 
I train on different nodes (VMs),  wonder for example, I train mnist.py example with 4 nodes. so each node will get the ENTIRE dataset and then extract what belongs to them? Or they will just get their own portion all from one GPU(rank 0 for example). Coz seems in the training loop, we pass the whole dateloader object for ecery GPU.
I am concern about this coz I have a very big dataset to run, if every VM needs to get the ENTIRE dataset, and then get their own, it will be very slow and storage demanding..

Sorry, I am kinda new to Horovod. Please kindly clarify that. Thanks!
